### DISCUSSION TITLE
Inconsistent behaviour with flutter stream

### DISCUSSION THREAD

**Discussion by angus-clark** [supabase/supabase#5256] [Questions] [answered]
I have implemented a realtime stream on one of my tables, however, sometimes the stream is not triggered on insert, update and/or delete. Often the stream will work initially (it will detect and insert event and update the frontend) but then subsequent events like another insert will do nothing.

I have RLS enabled on the table I am trying to stream with all policies set to authenticated users only. Also, I am streaming specific rows of the table.

I cannot find any solid answers/solutions from other discussions threads, and have tried toggling on and off the replication for the table in the supabase ui.

Are there any known bugs about this behavior or does someone have a solution?

SUMMARY: Realtime stream in flutter stops working/inconsistent behaviour, with no errors or indication why.

**Comment by GaryAustin1** [comment]
I can't comment directly on flutter, but make sure you have the most recent "drivers".  There have been several things fixed in the .js version that would have had to be moved into the flutter version POTENTIALLY. 
 
In the .js version I add this to the subscribe to see connection status and errors. 

.subscribe((status, e) => {
            console.log('status', status, e)
        })

 If you can do something similar in flutter/dart it might give you a clue on connection issues.

Normally you will lose connection on any internet interruption and potentially on the client going to low power mode (the JS browser environment in Chrome starts losing connection after 5 minutes as it throttles timers).  The JS client attempts to reconnect (I don't know about flutter/dart), but no updates during a disconnect will be sent, even later.  I have gone to handling the loss of connection by monitoring the subscription status and resetting everything (including a table update) on disconnect/reconnect.

There was also a token refresh issue (USUALLY AFTER 1 HOUR) that was fixed a month or so ago in realtime.js that sent it into a loop trying to reconnect with no error if you did not watch status.  No idea if flutter/dart even had the issue though.

Edit:
I glanced thru the dart repository and see several realtime open issues, (of course it is my slightly informed knowledge that dart is used with flutter) you probably want to post up there too, or look thru the issues.  With RLS changes to realtime several things had to be updated.

**Comment by w3b6x9** [comment]
@clarky2233 if you create a github issue in `realtime-dart` then one of the maintainers can address this: https://github.com/supabase-community/realtime-dart/issues/new

**Reply by angus-clark** [reply]
Thanks for your response. After further investigation the main issue I am having is that performing a delete operation does not trigger a real time event. My current implementation requires nested real time streams. I have tried to solve this with my own code using nested stream builders in flutter (a setup which worked with a firestore backend) and I have also followed the recommendations provided in this comment made by a contributor [https://github.com/supabase-community/supabase-dart/issues/44#issuecomment-903016255]().

Although the second approach seemed to resolve issues of the streams randomly closing, neither update the UI when a delete event is triggered. This includes when trying to delete from both the parent and child streams.

Is it possible to get some confirmation of this behavior by someone trying to reproduce the nested streams setup from the link I provided?

P.S. sorry for the vague nature of this issue

**Reply by GaryAustin1** [reply]
Edit.  Your link was just formatted badly.  Copying the text works.
The link you provided is to this discussion...

**Reply by GaryAustin1** [reply]
On your delete issue make sure you have done:
alter table "your_table" replica identity full;
Deletes won't come thru without that, pretty sure.

**Reply by angus-clark** [reply]
Sorry about the link not sure what happened there.

After setting the replica identity to full the child streams seem to be working nicely even on deletes.

However, for some reason the parent stream which generates the top level list has stopped working on all events.

This is my implementation of the code recommended in the link from early:

```
late StreamSubscription _athleteListener;

  final Map<String, StreamSubscription> _profileSubscriptions = {};

  final Map<String, List<AthleteData>> _athleteData = {};

  final List<Account> _selectedAthletes = [];

  void _loadData() async {
    _athleteListener = supabase
        .from(
            'training_data_accounts:training_data_id=eq.${widget.trainingDataID}')
        .stream(['id'])
        .order('added_date', ascending: true)
        .execute()
        .listen((athletes) async {
          for (final athlete in athletes) {
            Account account =
                await AccountService().getAccountFuture(athlete['account_id']);
            _selectedAthletes.add(account);

            // Make sure only create the listener if a listener for the user has not been created yet
            if (_profileSubscriptions[athlete['account_id']] == null) {
              _profileSubscriptions[athlete['account_id']] = supabase
                  .from(
                      'athlete_training_data:training_data_account_id=eq.${athlete['id']}')
                  .stream(['id'])
                  .execute()
                  .listen((event) {
                    _athleteData[athlete['account_id']] =
                        event.map((e) => AthleteData.fromJson(e)).toList();
                    setState(() {});
                  });
            }
            setState(() {});
          }
        });
  }

  @override
  void initState() {
    _loadData();
    super.initState();
  }

  @override
  void dispose() {
    _athleteListener.cancel();
    for (final profileListener in _profileSubscriptions.values) {
      profileListener.cancel();
    }
    super.dispose();
  }
```

**Reply by GaryAustin1** [reply]
The advice on needing 'alter table "your_table" replica identity full' was based on this:
[Improve documentation: DELETE events don't work for Realtime Row-Level Subscriptions while REPLICA IDENTITY is not FULL](https://github.com/supabase/realtime/issues/212)

And a quick read of Supabase Docs.
"_By default, Supabase does not send deleted records. If you want to receive the deleted record you can enable full replication for the table you are listening too:_ "

But latest realtime github docs say only use replica identity full if you want the entire contents of the deleted record (which is bad with RLS because delete changes ignore RLS for realtime.

I can confirm on latest Supabase instance the default replica identity setting allows deletes thru (with just record id).

@clarky2233 I don't know why then that made a difference for you...
@w3b6x9 The docs are somewhat confusing, not sure if this changed over time, or has been misconstrued in past answers.

Sorry for the sort of hijack of your discussion.

**Reply by angus-clark** [reply]
That's all good, I appreciate your help

**Reply by w3b6x9** [reply]
@clarky2233 Realtime RLS only works when tables have primary keys, and when they don't have keys, listening to those tables will not output anything. Setting replica identity to full will tell Postgres that the entire row becomes the key. I think that's why it's working when you did that.

Also, make sure that when you're listening to filters that the correct type is set as the value. For example, if I want to listen to a table where the column type is uuid and for the filter i set it as `111`, then Realtime RLS will crash and restart b/c it checks the type.

**Reply by gpbonillas** [reply]
Running the command: `alter table "table_name" replica identity full ` helped me

**Comment by chris311300** [comment, marked as answer]
@clarky2233 Hey, I had a similar issue. I found out that Realtime stops working if setState((){}) is called somewhere on the page after StreamBuilder(). I contacted supabase about the same. So, if you're passing the stream directly to the StreamBuilder(), realtime stops working after setState() as all widgets get rebuilt and refreshed. A possible fix could be, initializing the stream in the initState().

Something like this :

```
late final Stream<List<Map<String, dynamic>>> _stream;

  @override
  void initState() {
    _stream = Supabase.instance.client.from('my_table').stream(['id']).execute();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return StreamBuilder(
      stream: _stream,
      builder: (context, snapshot) {
        return Container();
      },
    );
  }

```

Hope this helps.

**Reply by leandrorazevedo** [reply]
This works great! But in my case, I need to use RLS with a value inside the JWT :

```
create policy "Allow only notifications to Owner's in JWT"
on notifications
for all
with check (
  (auth.jwt() ->> 'Id')::int = notifications."owner"
);
```

Basically, I use custom auth passing a JWT in the header:

```
await Supabase.initialize(
    url: 'xxxxx',
    anonKey: 'xxxxxx',
    headers: {
      'Authorization':
          'Bearer xxxxxxxxx',
    },
  );
```


RLS is working normally. But the STREAM stops working!



Can anyone help me?

**Comment by evanhzh** [comment]
My widget was not updating too but I removed initState and it worked for me. I implemented this with Riverpod, reading the stream provider in the build method. It seems like calling setState or initState would also stop the subscription. (Note: My stream subscription is in another file, outside of the widget and not initialized in initState).

**Reply by daytondang** [reply]
@evanzhoe I also use riverpod stream probider to watch the stream. Mine app is a webapp. On first run, the insert works and the frontend updated. But once I refresh the page, my stream doesn't work anymore... aka the insert does work for the backend, but the frontend doesn't update. How do you use your StreamProvider? With StreamBuilder? I use riverpod AsyncValue instead of StreamBuilder.

**Reply by gpbonillas** [reply]
Any idea, I have the same issue...

**Reply by HampusHauffman** [reply]
+1

**Reply by avalanche-tm** [reply]
I'm facing the same issue, with Riverpod, is there any update on this issue?

**Reply by angus-clark** [reply]
I am using Riverpod as well for my streams and everything seems to be working well for me so I thought I would outline how I have it set up:

Postgres Setup:
1. I created a realtime publication `create publciation realtime_supabase`
2. Added the table I want to stream to the publication `alter publication supabase_realtime add table my_table`
3. Set the replica identity to full `alter table my_table replica identity full`

Flutter/Riverpod Setup:
1. Defined a stream provider
```
final myStreamProvider = StreamProvider.autoDispose<>((ref) {
    return supabase
        .from('my_table')
        .stream(primaryKey: ['id'])
        .order('some_date_field');
});
```
Don't forget to specify your return type for the stream provider depending on your implementation.

2. Consume the stream provider in a widget
```
class MyStreamWidget extends ConsumerWidget {
  const MyStreamWidget({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final myStream = ref.watch(myStreamProvider);

    return myStream.when(
      data: (data) {
        // Add your widgets here
      },
      error: (err, stackTrace) {
        log(err.toString(), stackTrace: stackTrace);
        // Add your error widgets here
      },
      loading: () {
        // Add your loading widget here
      },
    );
  }
}
```

Hope this helps!

**Reply by evanhzh** [reply]
I found out what actually went wrong. I added **".eq"** filter after the stream method is called and that would not update the items in realtime after making changes to my backend.
It works after I removed the **equals** filter.

My sample code:


`  Stream<List<MyItem>> myItemStreamByCategory({required String tag}) =>
     _supabase.from('spending').stream(primaryKey: ['id'])       

// .eq('category_tag', tag)
          
.map((data) {
        return data.map((e) =>MyItem.fromJson(e)).toList();
      });
`

**Reply by arslanto** [reply]
@clarky2233 Thanks for your suggestion. Why did you not define the stream provider with async* ?

**Reply by angus-clark** [reply]
@tolunayar I believe no async with return, and async* with yield* do the same thing. At least I haven't noticed a different in behavior.

**Reply by FalconSlayer51** [reply]
Hey @clarky2233 i've encountered this problem with Supabase ,its just not working with StreamBuilder,the stream it provides when we listen to it also previous data, updated not been emitting, ive given my code below. can you suggest how to fix it.

```
class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  TextEditingController controller = TextEditingController();
  late Stream<List<Map<String, dynamic>>> _stream;

  @override
  void initState() {
    _stream = Supabase.instance.client.from('todo').stream(primaryKey: ['id']);
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: StreamBuilder<List<Map<String, dynamic>>>(
          stream: _stream,
          builder: (context, snapshot) {
            if (!snapshot.hasData) {
              return const CircularProgressIndicator();
            }
            var listOfData = snapshot.data;
            return ListView.builder(
              itemCount: listOfData!.length,
              itemBuilder: (context, index) {
                var item = listOfData[index];
                return ListTile(
                  key: Key(item['id'].toString()),
                  title: Text(item['name']),
                );
              },
            );
          },
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () async {
          await Supabase.instance.client.from('todo').insert({'name': 'hello'});
        },
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}
```

**Reply by angus-clark** [reply]
The only thing I can suggest is to double check replication is enabled in supabase for your table. You might need full identity replication.
